subsys/bluetooth/mesh/proxy_srv.c

```diff
@@ -899,250 +855,45 @@ static void gatt_connected(struct bt_conn *conn, uint8_t err)
-			       BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_NONE,
-			       NULL, NULL, NULL),
-	BT_GATT_CCC_MANAGED(&prov_ccc,
-			    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
-};
-
-static struct bt_gatt_service prov_svc = BT_GATT_SERVICE(prov_attrs);
-
-int bt_mesh_proxy_prov_enable(void)
-{
-	int i;
-
-	BT_DBG("");
-
-	if (gatt_svc == MESH_GATT_PROV) {
-		return -EALREADY;
-	}
-
-	if (gatt_svc != MESH_GATT_NONE) {
-		return -EBUSY;
-	}
-
-	(void)bt_gatt_service_register(&prov_svc);
-	gatt_svc = MESH_GATT_PROV;
-	prov_fast_adv = true;
-
-	for (i = 0; i < ARRAY_SIZE(clients); i++) {
-		if (clients[i].cli.conn) {
-			clients[i].filter_type = PROV;
-		}
-	}
-
-
-	return 0;
-}
-
-int bt_mesh_proxy_prov_disable(bool disconnect)
-{
-	int i;
-
-	BT_DBG("");
-
-	if (gatt_svc == MESH_GATT_NONE) {
-		return -EALREADY;
-	}
-
-	if (gatt_svc != MESH_GATT_PROV) {
-		return -EBUSY;
-	}
-
-	bt_gatt_service_unregister(&prov_svc);
-	gatt_svc = MESH_GATT_NONE;
-
-	for (i = 0; i < ARRAY_SIZE(clients); i++) {
-		struct bt_mesh_proxy_client *client = &clients[i];
-
-		if (!client->cli.conn || client->filter_type != PROV) {
-			continue;
-		}
-
-		if (disconnect) {
-			bt_conn_disconnect(client->cli.conn,
-					   BT_HCI_ERR_REMOTE_USER_TERM_CONN);
-		} else {
-			bt_mesh_pb_gatt_close(client->cli.conn);
-			client->filter_type = NONE;
-		}
-	}
-
-	bt_mesh_adv_update();
-
-	return 0;
-}
-static uint8_t prov_svc_data[20] = {
-	BT_UUID_16_ENCODE(BT_UUID_MESH_PROV_VAL),
-};
-
-static const struct bt_data prov_ad[] = {
-	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
-	BT_DATA_BYTES(BT_DATA_UUID16_ALL,
-		      BT_UUID_16_ENCODE(BT_UUID_MESH_PROV_VAL)),
-	BT_DATA(BT_DATA_SVC_DATA16, prov_svc_data, sizeof(prov_svc_data)),
-};
-
-int bt_mesh_pb_gatt_send(struct bt_conn *conn, struct net_buf_simple *buf,
-			 bt_gatt_complete_func_t end, void *user_data)
-{
-	struct bt_mesh_proxy_client *client = find_client(conn);
-
 	if (!client) {
-		BT_ERR("No Proxy Client found");
-		return -ENOTCONN;
-	}
-
-	if (client->filter_type != PROV) {
-		BT_ERR("Invalid PDU type for Proxy Client");
-		return -EINVAL;
-	}
-
-	return bt_mesh_proxy_msg_send(&client->cli, BT_MESH_PROXY_PROV, buf, end, user_data);
-}
-
-static size_t gatt_prov_adv_create(struct bt_data prov_sd[1])
-{
-	const struct bt_mesh_prov *prov = bt_mesh_prov_get();
-	size_t uri_len;
-
-	memcpy(prov_svc_data + 2, prov->uuid, 16);
-	sys_put_be16(prov->oob_info, prov_svc_data + 18);
-
-	if (!prov->uri) {
-		return 0;
-	}
-
-	uri_len = strlen(prov->uri);
-	if (uri_len > 29) {
-		/* There's no way to shorten an URI */
-		BT_WARN("Too long URI to fit advertising packet");
-		return 0;
+		BT_WARN("No Gatt Client found");
+		return;
 	}
 
-	prov_sd[0].type = BT_DATA_URI;
-	prov_sd[0].data_len = uri_len;
-	prov_sd[0].data = (const uint8_t *)prov->uri;
+	/* If this fails, the work handler exits early, as
+	 * there's no active connection.
+	 */
+	(void)k_work_cancel_delayable(&client->cli.sar_timer);
```

Comment:
Shouldn't this (and the corresponding `bt_mesh_proxy_msg_init`) be called in the pb_gatt.c file as well? I believe that requires proxy SAR too? 
If you follow the suggestion to move cli.buf into proxy_msg.c, we could probably move this too to that file.

Reply:
I'm think call this on gatt connect cb handler will be better, also will alloc buf in this function too.